gateway/src/endpoints/openai_compatible.rs

```diff
@@ -0,0 +1,1020 @@
+                Some(max_tokens.min(max_completion_tokens))
+            }
+            (Some(max_tokens), None) => Some(max_tokens),
+            (None, Some(max_completion_tokens)) => Some(max_completion_tokens),
+            (None, None) => None,
+        };
+        let input = openai_compatible_params.messages.try_into()?;
+        let chat_completion_inference_params = ChatCompletionInferenceParams {
+            temperature: openai_compatible_params.temperature,
+            max_tokens,
+            seed: openai_compatible_params.seed,
+            top_p: openai_compatible_params.top_p,
+            presence_penalty: openai_compatible_params.presence_penalty,
+            frequency_penalty: openai_compatible_params.frequency_penalty,
+        };
+        let inference_params = InferenceParams {
+            chat_completion: chat_completion_inference_params,
+        };
+        let variant_name = headers
+            .get("variant_name")
+            .map(|h| {
+                h.to_str()
+                    .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                        message: "variant_name header is not valid UTF-8".to_string(),
+                    })
+                    .map(|s| s.to_string())
+            })
+            .transpose()?;
+        let dryrun = headers
+            .get("dryrun")
+            .map(|h| {
+                h.to_str()
+                    .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                        message: "dryrun header is not valid UTF-8".to_string(),
+                    })
+                    .and_then(|s| {
+                        s.parse::<bool>()
+                            .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                                message: "dryrun header is not a valid boolean".to_string(),
+                            })
+                    })
+            })
+            .transpose()?;
+        let dynamic_tool_params = DynamicToolParams {
+            allowed_tools: None,
+            additional_tools: openai_compatible_params
+                .tools
+                .map(|tools| tools.into_iter().map(|tool| tool.into()).collect()),
+            tool_choice: openai_compatible_params
+                .tool_choice
+                .map(|tool_choice| tool_choice.into()),
+            parallel_tool_calls: openai_compatible_params.parallel_tool_calls,
+        };
+        let output_schema = match openai_compatible_params.response_format {
+            Some(OpenAICompatibleResponseFormat::JsonSchema { schema }) => Some(schema),
+            _ => None,
+        };
+        Ok(Params {
+            function_name,
+            episode_id,
+            input,
+            stream: openai_compatible_params.stream,
+            params: inference_params,
+            variant_name,
+            dryrun,
+            dynamic_tool_params,
+            output_schema,
+            credentials: InferenceCredentials::default(),
```

Comment:
Why not support this, at least for OpenAI? Or document we don't for now.

Reply:
I find it unintuitive to only support one kind of credential and it is not ergonomic to put more complicated JSON objects in headers. I will document that we don't for now

Reply:
we can snooze